examples/advanced/ocus_explanations.py

```diff
@@ -0,0 +1,302 @@
+
+            (2) sum x_l == 1 for l in (-Iend \ -I)
+                = exactly 1 literal explained at a time
+
+        Args
+        ----
+
+            hard (list[cpmpy.Expression]): Hard constraints
+
+            Iend (set): Cautious consequence, the set of literals that hold in
+                        all models.
+
+            I (set): partial interpretation (subset of Iend).
+
+        [1] Gamba, E., Bogaerts, B., & Guns, T. (8 2021). Efficiently Explaining CSPs
+        with Unsatisfiable Subset Optimization. In Z.-H. Zhou (Red), Proceedings of the
+        Thirtieth International Joint Conference on Artificial Intelligence,
+        IJCAI-21 (bll 1381–1388). doi:10.24963/ijcai.2021/191.
+
+        [2] Liffiton, M. H., & Sakallah, K. A. (2008). Algorithms for computing minimal
+        unsatisfiable subsets of constraints. Journal of Automated Reasoning, 40(1), 1-33.
+
+        [3] Reiter, R. (1987). A theory of diagnosis from first principles.
+        Artificial intelligence, 32(1), 57-95.
+    """
+
+    sat = SolverLookup.lookup(solver)(Model(hard))
+
+    ## OPT Model: Variables
+    hs_vars = boolvar(shape=len(Iend))
+    F = I | set(~var for var in Iend - I)
+
+    # id of variables that need to be explained
+    remaining_hs_vars = hs_vars[[id for id, var in enumerate(F) if var not in I]]
+
+    # mapping between hitting set variables hs_var <-> Iend
+    hs_vars_to_Iend = dict( (hs_vars[id], var) for id, var in enumerate(F))
+    Iend_to_hs_vars = dict( (var, hs_vars[id]) for id, var in enumerate(F))
+
+
+    # CONDITIONAL OPTIMISATION MODEL
+    hs_model = Model(
+        # exactly one variable to explain!
+        sum(remaining_hs_vars) == 1,
+        # optimal hitting set
+        minimize=sum(hs_var * cost(hs_vars_to_Iend[hs_var]) for hs_var in hs_vars)
+    )
+
+    ## instantiate hitting set solver
+    hittingset_solver = SolverLookup.lookup(solver)(hs_model)
+
+    while(True):
+        hittingset_solver.solve()
+
+        # Get hitting set
+        hs = hs_vars[hs_vars.value() == 1]
+
+        # map to vars of formula F
+        S = set(hs_vars_to_Iend[hs_var] for hs_var in hs)
+
+        if verbose > 0:
+            print("\n\t hs =", hs, S)
+
+        # SAT check and computation of model
+        if not sat.solve(assumptions=S):
+            if verbose > 0:
+                print("\n\t ===> OCUS =", S)
+
+            # deleting the hitting set solver
+            return S
+        # satisfying model
+        S = set(v == v.value() for v in F)
+
+        # compute complement of model in formula F
+        C =  F - S
+        set_to_hit = set(Iend_to_hs_vars[Iend_var] for Iend_var in F - S)
+
+        # Add complement as a new set to hit: sum x[j] * hij >= 1
+        hittingset_solver += (sum(set_to_hit) >= 1)
+
+        if verbose > 0:
+            print("\t Complement =", C)
+            print("\t set-to-hit =", set_to_hit)
+
+def optimal_propagate(hard, soft, solver="ortools", verbose=False):
```

Comment:
You are always solving with assump=soft, so it does not actually need the 'soft' formulation.

The function should also be called differently in case of integer variables...
what about 'solution_intersection' or smth? Note how you mention CNF in the comments, not a restriction here.

`def solution_intersection(model, solver="ortools", verbose=False)` should be doable. In that case, code can also be simpler I think, something like model2 = Model(model.constraints); while model2.solve(): model2 += blocking_clause; intersect = ...'. There might also be a way to do the intersections better... basically, user_vars is an array, so you can get an array of values of the same size... and then, some indicator on top, not sure. Smth is a bit weird with your current sets.